hnn_core/parallel_backends.py

```diff
@@ -69,6 +81,220 @@ def _read_all_bytes(fd, chunk_size=4096):
     return all_data
 
 
+def _read_stdout(fd, mask):
+    """Read stdout fd until receiving the process simulation is complete
+
+    Returns
+    -------
+    completion_signal : str | None
+        If completed, return 'end_of_sim'. Else, return None.
+    """
+    data = _read_all_bytes(fd)
+    if len(data) > 0:
+        str_data = data.decode()
+        if str_data == 'end_of_sim':
+            return str_data
+        elif 'end_of_sim' in str_data:
+            sys.stdout.write(str_data.replace('end_of_sim', ''))
+            return 'end_of_sim'
+
+        # output from process includes newlines
+        sys.stdout.write(str_data)
+
+    return None
+
+
+def _read_stderr(fd, mask):
+    """Read stderr from fd until end of simulation signal is received.
+
+    Parameters
+    ----------
+    fd : file object
+        The file object.
+
+    Returns
+    -------
+    proc_data_bytes : str
+        The processed data bytes before @.
+    data_len : str | int
+        The data length found as in @end_of_data:42@
+        where 42 is the data length. Returned only if
+        @ is found in the data read from stderr.
+    """
+    data = _read_all_bytes(fd)
+    if len(data) > 0:
+        str_data = data.decode()
+        if '@' in str_data:
+            # extract the signal
+            signal_index_start = str_data.rfind('@end_of_data:')
+            signal_index_end = str_data.rfind('@')
+            if signal_index_start < 0 or signal_index_end < 0 or \
+                    signal_index_end <= signal_index_start:
+                raise ValueError("Invalid signal start (%d) or end (%d) " %
+                                 (signal_index_start, signal_index_end) +
+                                 "index")
+
+            # signal without '@' on either side
+            signal = str_data[signal_index_start + 1:signal_index_end]
+
+            # remove the signal from the stderr output
+            spliced_str_data = str_data[0:signal_index_start] + \
+                str_data[signal_index_end + 1:]
+
+            # add the output bytes and return the signal
+            proc_data_bytes = spliced_str_data.encode()
+
+            split_string = signal.split(':')
+            if len(split_string) > 1 and len(split_string[1]) > 0:
+                data_len = int(split_string[1])
+            else:
+                raise ValueError("Completion signal from child MPI process"
+                                 " did not contain data length.")
+
+            return proc_data_bytes, data_len
+
+    return data
+
+
+def run_subprocess(command, pickled_obj, timeout, *args, **kwargs):
```

Comment:
tried to make something more generic. Didn't completely succeed, will try more in future PRs